drivers/gpio/gpio_mcux_lpc.c

```diff
@@ -109,49 +144,226 @@ static int gpio_mcux_lpc_read(struct device *dev,
+					    data->pint_id[pin]);
+		data->pin_callback_enables &= ~BIT(pin);
+	} else {
+		for (i = 0U; i < 32; i++) {
+			if (data->pint_id[i] != NO_PINT_INT) {
+				PINT_DisableCallbackByIndex(config->pint_base,
+							    data->pint_id[i]);
+			}
+		}
+		data->pin_callback_enables = 0U;
+	}
+
+	return 0;
+}
+
+static void gpio_mcux_lpc_port_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct gpio_mcux_lpc_config *config = dev->config->config_info;
+	struct gpio_mcux_lpc_data *data = dev->driver_data;
+	u32_t enabled_int;
+	u32_t int_flags;
+	u32_t pin;
+
+	for (pin = 0; pin < 32; pin++) {
+		if (data->pint_id[pin] != NO_PINT_INT) {
+			int_flags = PINT_PinInterruptGetStatus(
+				config->pint_base, data->pint_id[pin]);
+			enabled_int =
+				(int_flags << pin) & data->pin_callback_enables;
+
+			gpio_fire_callbacks(&data->callbacks, dev, enabled_int);
+
+			PINT_PinInterruptClrStatus(config->pint_base,
+						   data->pint_id[pin]);
+		}
+	}
+}
+
 static int gpio_mcux_lpc_init(struct device *dev)
 {
 	const struct gpio_mcux_lpc_config *config = dev->config->config_info;
+	struct gpio_mcux_lpc_data *data = dev->driver_data;
+	int i;
 
 	CLOCK_EnableClock(config->clock_ip_name);
 
+	for (i = 0; i < 32; i++) {
+		data->pint_id[i] = NO_PINT_INT;
+	}
+
 	return 0;
 }
 
 static const struct gpio_driver_api gpio_mcux_lpc_driver_api = {
 	.config = gpio_mcux_lpc_configure,
 	.write = gpio_mcux_lpc_write,
 	.read = gpio_mcux_lpc_read,
+	.manage_callback = gpio_mcux_lpc_manage_cb,
+	.enable_callback = gpio_mcux_lpc_enable_cb,
+	.disable_callback = gpio_mcux_lpc_disable_cb,
 };
 
 #ifdef CONFIG_GPIO_MCUX_LPC_PORT0
+static int lpc_gpio_0_init(struct device *dev);
+
 static const struct gpio_mcux_lpc_config gpio_mcux_lpc_port0_config = {
 	.gpio_base = GPIO,
+	.pint_base = PINT, /* TODO: SECPINT issue #16330 */
 	.port_no = PORT0_IDX,
 	.clock_ip_name = kCLOCK_Gpio0,
 };
 
 static struct gpio_mcux_lpc_data gpio_mcux_lpc_port0_data;
 
-DEVICE_AND_API_INIT(gpio_mcux_lpc_port0, CONFIG_GPIO_MCUX_LPC_PORT0_NAME,
-		    gpio_mcux_lpc_init,
-		    &gpio_mcux_lpc_port0_data, &gpio_mcux_lpc_port0_config,
-		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+DEVICE_AND_API_INIT(gpio_mcux_lpc_port0, DT_INST_0_NXP_LPC_GPIO_LABEL,
+		    lpc_gpio_0_init, &gpio_mcux_lpc_port0_data,
+		    &gpio_mcux_lpc_port0_config, POST_KERNEL,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
 		    &gpio_mcux_lpc_driver_api);
+
+static int lpc_gpio_0_init(struct device *dev)
+{
+	const struct gpio_mcux_lpc_config *config = dev->config->config_info;
+	struct gpio_mcux_lpc_data *data = dev->driver_data;
+	u32_t port = config->port_no;
+
+	gpio_mcux_lpc_init(dev);
+
+	/* Connect trigger sources to PINT */
+	INPUTMUX_Init(INPUTMUX);
+#ifdef DT_INST_0_NXP_LPC_GPIO_IRQ_0
+	INPUTMUX_AttachSignal(
+		INPUTMUX, pint_idx,
+		PIN_TO_INPUT_MUX_CONNECTION(port, DT_ALIAS_SW0_GPIOS_PIN));
+	data->pint_id[DT_ALIAS_SW0_GPIOS_PIN] = pint_idx++;
```

Comment:
Using `DT_ALIAS_SW0_GPIOS_PIN` here means that gpio interrupts will only work for the aliased buttons/switches. It will not work in the general case.

Reply:
@MaureenHelm What do you mean by "It will not work in the general case."?
InputMux device will map one pin to one interrupt (manual 19.6.3/4).

Analyzing the way I allocate PINT association, I'm not happy with the way I did it.
I'd rather have a static dts allocation, or even better, deduce PINT allocation according to isr no.

Reply:
You're always using the SW0 pin. What if I want to use a sensor pin for data ready interrupts?

Reply:
@MaureenHelm okay, I will change SW to something generic

Reply:
@MaureenHelm made the changes required, please review it again.
Thank you for your reviews!